### ISSUE TITLE
Expand and improve inter-contract comunication

### ISSUE THREAD

**Issue by steinerkelvin** [Kindelia/Kindelia#208]
- [ ] #204
- [ ] #232
- [ ] #230

---

- contracts that can run statements?
  - e.g. contracts can own namespaces and deploy code inside them

---

Some other possibilities:

- monotype buffer interface
  - which KHVM structure could be used? 
     - just a plain U120 cons-list
     - U120-tuple cons-list
     - Braun trees?
     - something else?
- marked/trusted contracts allowed to pass raw terms / "trusted call"
  - how?
  - complexity seems not worth it

Related to #35 and #116.
cc @o-santi

**Comment by o-santi** [comment]
What about adding a way to pattern match an specific type?
For example, adding a `#` to the variable in the following rule:
```
(Counter {Add #n}) =
  ask x = (Take);
  ask (Save (+ x n))
  (Done #0)
```
would make it so that only numbers could be pattern matched. This would add a little bit of overhead to the information stored about variables, but it would make for a more expressive inter-contract comunication (as there's no fixed way to send information).

Another place where computation could get stuck is when expecting lambdas:
```
fun (MutState f) {
  ask x = (Take);
  ask (Save (f x));
  (Done #0)
} with {
  #0
}
```
and for that, we could also add the `@` sign in front of `f`, in order to match only lambda terms.

If we restrict inter-contract communication to, say, only U120-cons list, **how would contracts pass lambdas to each other?** Isn't there value to it? Wouldn't it be more expressive if we didn't limit which kinds of terms we pass to one another?

**Comment by o-santi** [comment]
Of course, if the lambda term has a different arity than the contract expects, it could also get stuck. If the arguments to this lambda term are not what you expect (ie. a number instead of a constructor) it could also get stuck.  Probably the smallest type-checker that would be useful is somekind of simply-typed-lambda type checker, where instead of a single kind `*`, it has 3 kinds of terms: numbers, constructors and lambdas. So contracts could have types like `(Num -> (Lam Num Num) -> (Ctr (Ctr Num))`

Of course, it would add overhead to all the language, so this needs to be more well thought.

**Comment by steinerkelvin** [comment]
Earlier iterations of the design — last year, even before the HVM — had a simple, inexpensive type system that solved much of it because it was a simple system with basically ADTs. But a big point of the new design with the HVM is we can now do very high-level stuff with a clear cost model, and (a) there are no clear answers to what type of system would be the Right One*; (b) even the Right One would also add a lot of complexity to the system.

P.S. I see you have some actual simple type-checking in mind. As I'm not knowledgeable of type systems, I'll leave that discussion for you and Victor while I focus on finding other ways.
